src/libsync/discovery.cpp

```diff
@@ -0,0 +1,1218 @@
+                postProcessLocalNew();
+            }
+        } else {
+            item->_instruction = CSYNC_INSTRUCTION_SYNC;
+            if (noServerEntry) {
+                // Special case! deleted on server, modified on client, the instruction is then NEW
+                item->_instruction = CSYNC_INSTRUCTION_NEW;
+            }
+            item->_direction = SyncFileItem::Up;
+            item->_checksumHeader.clear();
+            item->_size = localEntry.size;
+            item->_modtime = localEntry.modtime;
+            item->_previousSize = dbEntry._fileSize;
+            item->_previousModtime = dbEntry._modtime;
+            if (!_dirItem || _dirItem->_direction == SyncFileItem::Down) {
+                _childModified = true;
+            }
+
+            // Checksum comparison at this stage is only enabled for .eml files,
+            // check #4754 #4755
+            bool isEmlFile = path._original.endsWith(QLatin1String(".eml"), Qt::CaseInsensitive);
+            if (isEmlFile && dbEntry._fileSize == localEntry.size && !dbEntry._checksumHeader.isEmpty()) {
+                if (computeLocalChecksum(dbEntry._checksumHeader, _discoveryData->_localDir + path._local, item)
+                        && item->_checksumHeader == dbEntry._checksumHeader) {
+                    qCInfo(lcDisco) << "NOTE: Checksums are identical, file did not actually change: " << path._local;
+                    item->_instruction = CSYNC_INSTRUCTION_UPDATE_METADATA;
+                }
+            }
+        }
+    } else if (_queryLocal == ParentNotChanged && dbEntry.isValid()) {
+        if (noServerEntry) {
+            // Not modified locally (ParentNotChanged), bit not on the server:  Removed on the server.
+            item->_instruction = CSYNC_INSTRUCTION_REMOVE;
+            item->_direction = SyncFileItem::Down;
+        }
+    } else if (noServerEntry) {
+        // Not locally, not on the server. The entry is stale!
+        qCInfo(lcDisco) << "Stale DB entry";
+        return;
+    } else if (dbEntry._type == ItemTypeVirtualFile) {
+        // If the virtual file is removed, recreate it.
+        item->_instruction = CSYNC_INSTRUCTION_NEW;
+        item->_direction = SyncFileItem::Down;
+        item->_type = ItemTypeVirtualFile;
+        item->_file.append(_discoveryData->_syncOptions._virtualFileSuffix);
+    } else if (!serverModified) {
+        // Removed locally: also remove on the server.
+        if (_dirItem && _dirItem->_isRestoration && _dirItem->_instruction == CSYNC_INSTRUCTION_NEW) {
+            // Also restore everything
+            item->_instruction = CSYNC_INSTRUCTION_NEW;
+            item->_direction = SyncFileItem::Down;
+            item->_isRestoration = true;
+            item->_errorString = tr("Not allowed to remove, restoring");
+        } else if (!dbEntry._serverHasIgnoredFiles) {
+            item->_instruction = CSYNC_INSTRUCTION_REMOVE;
+            item->_direction = SyncFileItem::Up;
+        }
+    }
```

Comment:
Generally: I find these huge if-elseif-... chains very hard to follow. A lot of logic depends on what has previously been ruled out. Any ideas for making this more readable? Maybe asserting all the other important assumed conditions after the `if`?

Like, I've read this but:
- I'm not sure it's exhaustive / that all cases are covered.
- If there were mutually exclusive conditions I probably wouldn't have spotted them.
- Usually the branches assume a lot of context that's way up in this function.
- Maybe renames and conflict checking could be functions to limit the indent-depth and height of this thing?
- Overall feeling of fragility: due to many long-term assumptions and connections

Merging three trees in this way is genuinely hard and complex. Some of this will always be there.